user/src/com/google/gwt/user/server/rpc/RemoteServiceServlet.java

```diff
@@ -520,12 +544,9 @@ private void putCachedSerializationPolicy(String moduleBaseURL,
     }
   }
 
-  private void writeResponse(HttpServletRequest request,
-      HttpServletResponse response, String responsePayload) throws IOException {
-    boolean gzipEncode = RPCServletUtils.acceptsGzipEncoding(request)
-        && shouldCompressResponse(request, response, responsePayload);
```

Comment:
Worth calling out, dropping this changes behavior slightly, server spends more time gzipping too-small-to-care payloads, even if we happen to have the string length available. 

I wonder if we could pass the string length if known (or null check the payload and then check the length), but if null/unknown just go ahead and gzip it?

Reply:
It would probably require splitting between old and new version before the stream is assembled, and then for the old version first again assembling everything before computing its size and then deciding about gzipping. With all the array initialization and copying added by the "old" approach I wonder if it's amortizing the GZIPOutputStream init overhead, yet I'll try to measure that at least for our average workloads. I'll report back here.

Reply:
How about this: Currently the limit for gzipping is set to 256B, assuming that the payload doubles in size (content.length()*2 > UNCOMPRESSED_BYTE_SIZE_LIMIT) where with UTF-8 it probably doesn't quite. Still, we could wait for these first UNCOMPRESSED_BYTE_SIZE_LIMIT bytes and buffer them. If the content is complete before the limit is exceeded, the content can be written out uncompressed. If the size exceeds the limit, the header field for GZIP compression is set, then a GZIPOutputStream is created, the buffer so far is dumped into the GZIPOutputStream and all further content is streamed.

This will limit memory footprint to those additional 256B for the temporary buffer before the actual streaming kicks in.

I will work on a corresponding suggestion.

Reply:
f82eeab9039fe4124138bdd713dff066bbccb2b6 has a solution, along my previous comment: buffering the first UNCOMPRESSED_BYTE_SIZE_LIMIT bytes and simply flushing the buffer if its size isn't exceeded; otherwise activate compression, flush the buffer to the compressing output stream and continue streaming while releasing the buffer.

The tests you mention run fine so far. I'll add tests for the new version, too.

Reply:
Just out of curiosity:

`
  public void testVersion8Fallbacks() {
    StringBuilder longString = new StringBuilder(66000);
    for (int i = 0; i < 660000; i++) {
      longString.append("a");
    }
`
looks as if the for-loop was meant to only run up to 66000, not 660000. With the array split topic removed I think this test should probably be removed, too. Do you agree?